src/lxc/lxccontainer.c

```diff
@@ -1771,24 +1771,52 @@ static bool lxcapi_create(struct lxc_container *c, const char *t,
 
 static bool do_lxcapi_reboot(struct lxc_container *c)
 {
+	int ret;
 	pid_t pid;
-	int rebootsignal = SIGINT;
+	int rebootsignal = SIGINT, sock = -1;
+	bool retv = false;
+	lxc_state_t states[MAX_STATE] = {0};
 
 	if (!c)
 		return false;
+
 	if (!do_lxcapi_is_running(c))
 		return false;
+
 	pid = do_lxcapi_init_pid(c);
 	if (pid <= 0)
 		return false;
+
 	if (c->lxc_conf && c->lxc_conf->rebootsignal)
 		rebootsignal = c->lxc_conf->rebootsignal;
+
+	/* Add a new state client before sending the reboot signal so that we
+	 * don't miss a state.
+	 */
+	states[RUNNING] = 2;
+	ret = lxc_cmd_add_state_client(c->name, c->config_path, states, &sock);
+
 	if (kill(pid, rebootsignal) < 0) {
-		WARN("Could not send signal %d to pid %d.", rebootsignal, pid);
+		WARN("Could not send signal %d to pid %d", rebootsignal, pid);
+		close(sock);
 		return false;
 	}
-	return true;
 
+	/* Retrieve the state. */
+	if (sock >= 0) {
+		int state;
+		state = lxc_cmd_sock_rcv_state(sock, 0);
+		close(sock);
+		if (state != RUNNING)
+			return false;
+		TRACE("Container rebooted");
+		retv = true;
+	} else {
+		TRACE("Received state \"%s\" instead of expected \"RUNNING\"",
+		      lxc_state2str(ret));
+	}
```

Comment:
Hmm, so am I missing something or is this turning reboot() from a non-blocking API call to a blocking API call?

Unlike shutdown() there is no timeout argument here for the caller to indicate that they just want the signal sent but not wait for something to actually happen.

Reply:
So probably better to add a `reboot2()` that does both jobs. API-wise it'll be pretty nice to finally having a blocking reboot call...

Reply:
Yeah, I'd be fine with keeping reboot() be just signal sending and reboot2() be signal sending + wait with a timeout.